Test failures on Linux/i386

I noticed that two tests were failing when building the Debian goawk package inside a i386 schroot.

```
?   	github.com/benhoyt/goawk/scripts/csvbench/count	[no test files]
?   	github.com/benhoyt/goawk/scripts/csvbench/write	[no test files]
--- FAIL: TestAWK (0.29s)
    --- FAIL: TestAWK/p.50 (0.01s)
        goawk_test.go:147: output differs, run: git diff testdata/output/p.50
FAIL
FAIL	github.com/benhoyt/goawk	0.832s
ok  	github.com/benhoyt/goawk/internal/ast	0.003s
ok  	github.com/benhoyt/goawk/internal/compiler	0.005s
ok  	github.com/benhoyt/goawk/internal/cover	0.003s
ok  	github.com/benhoyt/goawk/internal/parseutil	0.003s
ok  	github.com/benhoyt/goawk/internal/resolver	0.004s
--- FAIL: TestInterp (0.64s)
    --- FAIL: TestInterp/{_print_$1+0_}__#_+posix (0.00s)
        interp_test.go:1223: expected/got:
            "0\n16\n78187493520\n11259375\n11259375\n-10\n10\n15.9375\n16320\n0.015564\n0.9375\n15\n0\n"
            "0\n16\n7.81875e+10\n11259375\n11259375\n-10\n10\n15.9375\n16320\n0.015564\n0.9375\n15\n0\n"
--- FAIL: TestNative (0.00s)
    --- FAIL: TestNative/_BEGIN_{__print__print_bool(),_bool(0),_bool(1),_bool(""),_bool("0"),_ (0.00s)
        interp_test.go:1223: expected/got:
            "\n0 0 1 0 1 1\n0 42 -5 3 -3\n0 42 -5 127 -128\n0 42 -5 32767 -32768\n0 42 -5 2147483647 -2147483648\n0 42 -5 2147483647000 -2147483647000\n0 42 0 1\n0 42 251 127 128 255\n0 42 65535 65535\n0 42 4294967295 4294967295\n0 42 1 4294967296 2147483647000\n..Foo bar.1234\n..Foo bar.1234\n"
            "\n0 0 1 0 1 1\n0 42 -5 3 -3\n0 42 -5 127 -128\n0 42 -5 32767 -32768\n0 42 -5 2147483647 -2147483648\n0 42 -5 2.14748e+12 -2.14748e+12\n0 42 0 1\n0 42 251 127 128 255\n0 42 65535 65535\n0 42 4.29497e+09 4.29497e+09\n0 42 1 4.29497e+09 2.14748e+12\n..Foo bar.1234\n..Foo bar.1234\n"
FAIL
FAIL	github.com/benhoyt/goawk/interp	0.708s
ok  	github.com/benhoyt/goawk/lexer	0.003s
ok  	github.com/benhoyt/goawk/parser	0.124s
FAIL
``` 


I initially thought this was a Debian-specific problem, but I can also reproduce it on alpine.

```bash
docker run --rm -v $(pwd):/goawk i386/golang:bookworm \
  sh -c "cd /goawk && apt-get update && apt-get install -y gawk procps && go test ./..."
```

```bash
docker run --rm -v $(pwd):/goawk i386/golang:alpine \
  sh -c "cd /goawk && pkg update && pkg add gawk && go test ./..."
```

The failures are related to number formatting. This is not a big deal, and I can probably skip them when building for non-64bit architectures, but maybe it's worth fixing upstream :^)

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -29,6 +29,7 @@ jobs:
       run: |
         gawk --version
         go test -race ./...
+        GOARCH=386 go test ./...
 
   build-windows:
     runs-on: windows-latest

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -29,7 +29,6 @@ jobs:
       run: |
         gawk --version
         go test -race ./...
-        go test ./...
 
   build-linux-386:
     runs-on: ubuntu-latest

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -52,7 +52,6 @@ jobs:
     - name: Build and Test
       run: |
         gawk --version
-        GOARCH=386 go test -race ./...
         GOARCH=386 go test ./...
 
   build-windows:

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -29,6 +29,30 @@ jobs:
       run: |
         gawk --version
         go test -race ./...
+        go test ./...
+
+  build-linux-386:
+    runs-on: ubuntu-latest
+
+    strategy:
+      fail-fast: false
+      matrix:
+        go: ['1.22']
+
+    name: Go ${{ matrix.go }} on Linux 386
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Set up Go
+      uses: actions/setup-go@v2
+      with:
+        go-version: ${{ matrix.go }}
+
+    - name: Build and Test
+      run: |
+        gawk --version
+        GOARCH=386 go test -race ./...
         GOARCH=386 go test ./...
 
   build-windows:

--- a/internal/ast/ast.go
+++ b/internal/ast/ast.go
@@ -285,8 +285,8 @@ type NumExpr struct {
 }
 
 func (e *NumExpr) String() string {
-	if e.Value == float64(int(e.Value)) {
-		return strconv.Itoa(int(e.Value))
+	if e.Value == float64(int64(e.Value)) {
+		return strconv.FormatInt(int64(e.Value), 10)
 	} else {
 		return fmt.Sprintf("%.6g", e.Value)
 	}

--- a/internal/compiler/compiler.go
+++ b/internal/compiler/compiler.go
@@ -1135,10 +1135,10 @@ func (c *compiler) binaryOp(op lexer.Token) {
 // Generate an array index, handling multi-indexes properly.
 func (c *compiler) index(index []ast.Expr) {
 	for _, expr := range index {
-		if e, ok := expr.(*ast.NumExpr); ok && e.Value == float64(int(e.Value)) {
+		if e, ok := expr.(*ast.NumExpr); ok && e.Value == float64(int64(e.Value)) {
 			// If index expression is integer constant, optimize to string "n"
 			// to avoid toString() at runtime.
-			s := strconv.Itoa(int(e.Value))
+			s := strconv.FormatInt(int64(e.Value), 10)
 			c.expr(&ast.StrExpr{Value: s})
 			continue
 		}

--- a/internal/compiler/disassembler.go
+++ b/internal/compiler/disassembler.go
@@ -134,8 +134,8 @@ func (d *disassembler) disassemble(prefix string) error {
 		case Num:
 			index := d.fetch()
 			num := d.program.Nums[index]
-			if num == float64(int(num)) {
-				d.writeOpf("Num %d (%d)", int(num), index)
+			if num == float64(int64(num)) {
+				d.writeOpf("Num %d (%d)", int64(num), index)
 			} else {
 				d.writeOpf("Num %.6g (%d)", num, index)
 			}

--- a/interp/functions.go
+++ b/interp/functions.go
@@ -405,11 +405,11 @@ func (p *interp) sprintf(format string, args []value) (string, error) {
 		case 's':
 			v = p.toString(a)
 		case 'd':
-			v = int(a.num())
+			v = int64(a.num())
 		case 'f':
 			v = a.num()
 		case 'u':
-			v = uint(a.num())
+			v = uint64(a.num())
 		case 'c':
 			var c []byte
 			n, isStr := a.isTrueStr()

--- a/interp/value.go
+++ b/interp/value.go
@@ -141,8 +141,8 @@ func (v value) str(floatFormat string) string {
 			} else {
 				return "inf"
 			}
-		case v.n == float64(int(v.n)):
-			return strconv.Itoa(int(v.n))
+		case v.n == float64(int64(v.n)):
+			return strconv.FormatInt(int64(v.n), 10)
 		default:
 			if floatFormat == "%.6g" {
 				return strconv.FormatFloat(v.n, 'g', 6, 64)

--- a/interp/vm.go
+++ b/interp/vm.go
@@ -980,7 +980,7 @@ func (p *interp) callBuiltin(builtinOp compiler.BuiltinOp) error {
 		p.replaceTop(num(float64(index + 1)))
 
 	case compiler.BuiltinInt:
-		p.replaceTop(num(float64(int(p.peekTop().num()))))
+		p.replaceTop(num(float64(int64(p.peekTop().num()))))
 
 	case compiler.BuiltinLength:
 		p.push(num(float64(len(p.line))))

